### ISSUE TITLE
share: Cache Availability

### ISSUE THREAD

**Issue by Wondertan** [celestiaorg/celestia-node#180]
## Background
We have defined `share.Availability` interface which validates the availability of shares for a specific `share.Root`. In case we call `SharesAvailable` multiple times over the same Root, we will execute availability again. Instead, we should not execute expensive validation and perform caching of successful and failed executions. This way Light Availability won't sample shares again over the same Root. 

## Design
To fix the problem above we should introduce another implementation of `share.Availability` which wraps e.g Light Availability and caches the results on disk.

Basically, on `SharesAvailable` Cache Availability checks if there is a cached result on disk for a given Root and if not pass the call to `SharesAvailable` of wrapped `share.Availability`. Once done, Cache Availability saves result on disk for the Root. Further, If someone calls `SharesAvailable` Cache Availability  will return result from cache, without re-executing `SharesAvailable` on a wrapped `share.Availability`.

**Comment by renaynay** [comment]
<img width="824" alt="Screenshot 2022-03-21 at 18 49 55" src="https://user-images.githubusercontent.com/41963722/159333699-6840d1e9-9479-4514-89d5-139187f467cb.png">

- @Wondertan comment on #473
